Pocket/recommendation-api
# feat: Emit scheduled_surface_item_id to Snowplow

# Goal
Emit `scheduled_surface_item_id` to Snowplow for each New Tab recommendation, to allow analysts to join directly between a [recommendation event](https://console.snowplowanalytics.com/cf0fba6b-23b3-49a0-9d79-7ce18b8f9618/data-structures/f8e9cc9ff75a5ac53276a511523b7b61bd029210ef38996d879b02f1d20da465) and [corpus schedule event](https://console.snowplowanalytics.com/cf0fba6b-23b3-49a0-9d79-7ce18b8f9618/data-structures/c8fa3b61910abeecee71229dfab344af3deceb1dc4f51bd44f642ae5c60282f0).

Currently, this join is ambiguous because it is made using the CorpusItem id, and CorpusItems can be scheduled mutliple times. This results in incorrect reporting when items are scheduled multiple times. [MC-342](https://mozilla-hub.atlassian.net/browse/MC-342)

## How to check this is working
1. Start recommendation-api using the 'main Pocket-Dev' run configuration
2. Make a local request to recommendation-api using the Postman RecommendationAPI/newTabSlate request
3. Open [Snowplow Mini Kibana filtered on app_id pocket-data-products-recommendation-api-development](https://com-getpocket-prod1.mini.snplow.net/kibana/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-15m,to:now))&_a=(columns:!(_source),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:good,key:app_id,negate:!f,params:(query:pocket-data-products-recommendation-api-development),type:phrase),query:(match_phrase:(app_id:pocket-data-products-recommendation-api-development)))),index:good,interval:auto,query:(language:kuery,query:''),sort:!()))

Check that the corpus_slate context has `scheduled_surface_item_id` values for each recommendation:

![image](https://github.com/Pocket/recommendation-api/assets/1547251/0c011fc4-2777-4591-8b0a-9efa818a1e2b)


## Deployment
1. - [x] Wait until code freeze is over on Jan 2nd, 11pm
5. - [x] Deploy [corpus_slate](https://console.snowplowanalytics.com/cf0fba6b-23b3-49a0-9d79-7ce18b8f9618/data-structures/211240b158a39edba9bfa155e221af4ad8fe04eaec09ad968e54ffe659e07dc1) in Snowplow https://github.com/Pocket/spec/pull/202
6. - [ ] Deploy recommendation-api https://github.com/Pocket/recommendation-api/pull/1074
7. - [ ] Deploy dbt-snowflake changes

## Reference

Tickets:
* https://mozilla-hub.atlassian.net/browse/MC-342

Snowplow:
* [corpus_slate entity](https://console.snowplowanalytics.com/cf0fba6b-23b3-49a0-9d79-7ce18b8f9618/data-structures/211240b158a39edba9bfa155e221af4ad8fe04eaec09ad968e54ffe659e07dc1)

[MC-342]: https://mozilla-hub.atlassian.net/browse/MC-342?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
